### PR TITLE
fix: allow listener removal via DataChangeEvent

### LINKED ISSUE #6947: DataChangeEvent can't unregister its listener
`ComponentEvent`s have a useful `unregisterListener` method to be able to unregister themselves once invoked, but `DataChangeEvent` does not; it would be useful to set a "once" listener (not sure if there's another way to do it).

### PR DESCRIPTION
**Pull request by denis-anisimov**
fixes #6947

## Description

Please list all relevant dependencies in this section and provide summary of the change, motivation and context.

Fixes # (issue)

## Type of change

- [ ] Bugfix
- [x] Feature

## Checklist

- [ ] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [ ] I have added a description following the guideline.
- [ ] The issue is created in the corresponding repository and I have referenced it.
- [ ] I have added tests to ensure my change is effective and works as intended.
- [ ] New and existing tests are passing locally with my change.
- [ ] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [ ] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

File: flow-data/src/main/java/com/vaadin/flow/data/provider/DataChangeEvent.java [hunk 1]
```diff
@@ -112,4 +116,31 @@ public DataChangeEvent(DataProvider<T, ?> source) {
...
+     *            the unregister command
+     */
+    void setUnregisterListenerCommand(Command unregisterListenerCommand) {
+        this.unregisterListenerCommand = unregisterListenerCommand;
+    }
+
+    /**
+     * Unregisters the event listener currently being invoked.
+     * <p>
+     * This method can only be called from within an event listener. Calling it
```
Comment (mshabarov) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/DataChangeEvent.java:135]: "can only be called" doesn't explain clearly what happens if devs do call it outside the listener, so I'd propose to reflect in javadoc that this method throws when called outside of the listener.

File: flow-data/src/test/java/com/vaadin/flow/data/provider/AbstractDataProviderTest.java [hunk 2]
```diff
@@ -75,4 +73,30 @@ public void removeListener_listenerIsNotNotified() {
         dataProvider.refreshAll();
         Assert.assertNull(event.get());
     }
+
+    @Test
+    public void eventUnregisterListener_insideListener() {
```
Comment (mshabarov) [inline flow-data/src/test/java/com/vaadin/flow/data/provider/AbstractDataProviderTest.java:78]: Add test `eventUnregisterListener_outsideListener()`

Comment (mshabarov) [reply flow-data/src/test/java/com/vaadin/flow/data/provider/AbstractDataProviderTest.java:78]: I'd add one more test which adds a listener which throws an exception, and then check that unregister listener command in `null` despite of the exception thrown.

**Review by mshabarov — CHANGES_REQUESTED** [review]

File: flow-data/src/main/java/com/vaadin/flow/data/provider/DataChangeEvent.java (hunk 1 above)
Comment (denis-anisimov) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/DataChangeEvent.java:135]: Copied from here : https://github.com/vaadin/flow/blob/b6cf0b5ea13f9a7557853f480b7e49806d40f387/flow-server/src/main/java/com/vaadin/flow/component/ComponentEvent.java#L84

Would be better to have them consistent....
But OK.

File: flow-data/src/test/java/com/vaadin/flow/data/provider/AbstractDataProviderTest.java (hunk 2 above)
Comment (denis-anisimov) [reply flow-data/src/test/java/com/vaadin/flow/data/provider/AbstractDataProviderTest.java:78]: done and done

**Review by mshabarov — APPROVED** [review]

**Pull request merged**